### ISSUE TITLE
Technical Overview on How to Reduce the Index Size

### ISSUE THREAD

**Issue by Kerollmops** [meilisearch/meilisearch#3761]
This issue summarizes the different ways we can reduce the size of an index. We list low-hanging fruits and much harder-to-catch fruits too.

### Get Rid of the WordDocids and ExactWordDocids

With the new search refactor, we introduced multiple new databases that store the docids by field instead of global, but we kept the global inverted index, too, for performance reasons. We could eliminate both global inverted indices, do more unions at runtime, and ensure acceptable performances.

### Skip Indexing Some Fields

We could detect and not index some fields we could identify as only containing. We can add those fields to a `nonSearchableAttributes` setting:
  - Document ID
  - numbers
  - URLs
  - UUIDs
  - Date and Time

Doing so would be breaking for many use cases but will also help reduce the size of the FST and the number of entries in the different databases.

### Improve the Serialization of the RoaringBimaps

We use the roaring bitmaps data structure a lot in the engine. Nearly all of the databases of Meilisearch use roaring bitmaps as their values. By introducing a smaller (and faster) serialization format, we could reduce the size of an index.

We already use a custom format that directly writes the integers when it weighs less than the default serialization header. If it weighs more, we use the default format that uses a header.

### Ignore Bucketed Word Positions that are Far

We could store only the first (bucketed) position of a word instead of all of them -> this will negatively impact the ExactAttribute ranking rule so that we could make an exception for the first ~16 positions.

TBD Loïc

### Bucket Word Positions more Aggressively

We could bucket the positions even more aggressively.

TBD Loïc

### Get Rid of the Soft Deleted Documents

The search algorithm no longer uses some of our databases, as we now only have one single algorithm. Before the new search refactoring, we decided between two ways of computing the documents based on the number of candidates to find.

Unfortunately, we use those databases when we are deleting documents. The reason is that deleting documents can take a lot of time because we need to know which entry in the inverted indexes must be updated (remove the deleted document ids).

We introduced the soft-deleted documents to postpone this operation only once in a while and hide the fact that it costs a lot. This new system is hard to maintain as soft-deleted document ids are not _polluting_ all of our databases, and we must always ensure we are not returning deleted documents to the user.

We recently thought we could drastically speed up the deletion operation by reindexing the documents to gather the entries we must update and avoid iterating over the whole database. Therefore, we could remove those direct values databases, e.g., `document id -> facet values`, save space, and simplify the engine even more.

### Identify Words by Using Numbers

We could store words as variable-length integers in the DB (ouch…)
As a kind of string interning. I’ve checked the average size of words, and I think it was ~5 / 6 bytes. But using integers, we could reduce that to ~2 bytes for most of them.

TBD Loïc

### Remove the DocidWordPosition Database

We no more use the `docid_word_positions` database. We can remove it as soon as possible. Doing so will reduce the indexing time and the on-disk size of the database.

### Limit the Maximum Proximity

The current maximum proximity is 8, and we can reduce it to 4. It would divide by nearly half the number of entries in the database but could impact the relevancy.

### Disable Proximity

We could be a little smarter when deciding which database must be indexed and, therefore, not index anything related to the proximity databases when the `Proximity` ranking rule is missing from the settings.

One counterpoint is that we can’t remove the proximity indexing entirely because we use it for phrase searches.

### Reformat the WordFidDocids Database

We could reformat the word_fid_docids database to go from `(word, fid) -> docids` to `word -> [(fid, docids)]` (same change as for the proximity database). This way, instead of duplicating the word for every field containing it, we only store it once.

However, we must be careful as we will store the docids roaring bitmaps in the value, and values are limited to ~2GiB to ~4GiB in LMDB.

### Reformat the WordPairProximityDocids Database

Changing the format of the proximity database is the biggest win to me (with no relevancy cost) + it will speed up search requests as well.

TBD Loïc

### Reformat the Documents database

In some use cases with many small documents, we measured that the documents database is one of the biggest. We could compress the raw documents (in grenad files of less than 2GiB) instead of keeping them raw in LMDB. We keep raw data when access time is important. However, fetching 20 documents in a compressed grenad at the end of a search should be fast enough.

One counterpoint is that the index sometimes weighs as many as 20x documents, so compressing the document database won’t significantly impact the size of the database and can complexify the use of the engine.
